### PR TITLE
Add support for improved error reporting.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This adds support for improved error reporting:
 * Typing errors have additional explanation added.
 * Errors in general now have source line information and a caret
   pointing to the location of the problem. Tracebacks through the
   compiler are also now hidden by default.
 * All errors now have a help message attached as follows:
    * Typing errors link to the supported functionality and debugging
      documentation.
    * New error class "UnsupportedError" is raised when something that
      is likely unsupported is encountered. This links to github to
      put in a feature request.
    * All other errors ask users to file bug reports and refer to the
      github issue tracker.
 * There are 3 new environment flags:
    * One to control showing full tracebacks right through the compiler.
    * One to toggle showing help.
    * One to switch to developer mode which enables full tracebacks and
      disables showing the help information.
 * If the `colorama` package is installed the terminal output is
   highlighted when a compatible exception is raised.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
@seibert @sklam @ehsantn @DrTodd13  I've been doing development rebased on this branch for quite a few weeks now and I'm getting to the point where I'm almost happy with what it does. So I'm not going to do any more without feedback/input from others as this sort of change is somewhat subjective. Please do try developing against this branch if you wish, or just write some code that'll fail/fish it out of the issue tracker and see what you think of the error formatting and new behaviour. If you install `colorama` it should enable highlighting too.

I do not expect the tests to pass (probably error message string parse/match problems).

File: numba/errors.py [hunk 1]
```diff
@@ -110,11 +335,12 @@ def contexts(self):
 
     def add_context(self, msg):
         """
-        Add contextual info.  The execption message is expanded with the new
-        contextual info.
+        Add contextual info.  The exeception message is expanded with the new
```
Comment (seibert) [inline numba/errors.py]: "exception"

**Comment by seibert** [comment]
This looks great!  (I had one spelling nitpick.)  I think this will be a huge improvement, and probably the best thing is to ensure this makes it into the RC so we can get user feedback.

**Comment by seibert** [comment]
Found a Py27 issue running inside IPython:
``` python
Python 2.7.12 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:43:17)
Type "copyright", "credits" or "license" for more information.

IPython 5.4.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from numba import njit
   ...: @njit
   ...: def foo(x):
   ...:     return [x] + x
   ...:
   ...: foo(1)
   ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-e085d24e1e94> in <module>()
      4     return [x] + x
      5
----> 6 foo(1)

/Users/sseibert/continuum/numba/numba/dispatcher.pyc in _compile_for_args(self, *args, **kws)
    354                     e.patch_message(''.join(e.args) + help_msg)
    355             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 356             raise e
    357
    358     def inspect_llvm(self, signature=None):

TypeError: Caused By:
Traceback (most recent call last):
  File "numba/compiler.py", line 241, in run
    stage()
  File "numba/compiler.py", line 455, in stage_nopython_frontend
    self.locals)
  File "numba/compiler.py", line 882, in type_inference_stage
    infer.propagate()
  File "numba/typeinfer.py", line 847, in propagate
    errors = self.constraints.propagate(self)
  File "numba/typeinfer.py", line 142, in propagate
    highlighting=False)
  File "numba/errors.py", line 324, in __init__
    highlight("%s\n%s\n" % (msg, loc.strformat())))
  File "numba/ir.py", line 88, in strformat
    spaces = count_spaces(selected[-1])
  File "numba/ir.py", line 60, in count_spaces
    for x in itertools.takewhile(str.isspace, string):
TypeError: descriptor 'isspace' requires a 'str' object but received a 'unicode'

Failed at nopython (nopython frontend)
descriptor 'isspace' requires a 'str' object but received a 'unicode'
```

**Comment by seibert** [comment]
Also, when I run the same code at the command line, should I be surprised that I see the list of types printed twice?
```
(base) 0179-sseibert:numba sseibert$ python foo.py
Traceback (most recent call last):
  File "foo.py", line 6, in <module>
    foo(1)
  File "/Users/sseibert/continuum/numba/numba/dispatcher.py", line 340, in _compile_for_args
    reraise(type(e), e, None)
  File "<string>", line 2, in reraise
numba.errors.TypingError: Caused By:
Traceback (most recent call last):
  File "/Users/sseibert/continuum/numba/numba/compiler.py", line 241, in run
    stage()
  File "/Users/sseibert/continuum/numba/numba/compiler.py", line 455, in stage_nopython_frontend
    self.locals)
  File "/Users/sseibert/continuum/numba/numba/compiler.py", line 882, in type_inference_stage
    infer.propagate()
  File "/Users/sseibert/continuum/numba/numba/typeinfer.py", line 852, in propagate
    raise errors[0]
TypingError: Invalid usage of + with parameters (list(int64), int64)
Known signatures:
 * (int64, int64) -> int64
 * (int64, uint64) -> int64
 * (uint64, int64) -> int64
 * (uint64, uint64) -> uint64
 * (float32, float32) -> float32
 * (float64, float64) -> float64
 * (complex64, complex64) -> complex64
 * (complex128, complex128) -> complex128
 * (uint64,) -> uint64
 * (uint16,) -> uint64
 * (uint8,) -> uint64
 * (uint32,) -> uint64
 * (int32,) -> int64
 * (int16,) -> int64
 * (int64,) -> int64
 * (int8,) -> int64
 * (float32,) -> float32
 * (float64,) -> float64
 * (complex64,) -> complex64
 * (complex128,) -> complex128
 * parameterized
[1] During: typing of intrinsic-call at foo.py (4)

File "foo.py", line 4:
def foo(x):
    return [x] + x
    ^


Failed at nopython (nopython frontend)
Invalid usage of + with parameters (list(int64), int64)
Known signatures:
 * (int64, int64) -> int64
 * (int64, uint64) -> int64
 * (uint64, int64) -> int64
 * (uint64, uint64) -> uint64
 * (float32, float32) -> float32
 * (float64, float64) -> float64
 * (complex64, complex64) -> complex64
 * (complex128, complex128) -> complex128
 * (uint64,) -> uint64
 * (uint16,) -> uint64
 * (uint8,) -> uint64
 * (uint32,) -> uint64
 * (int32,) -> int64
 * (int16,) -> int64
 * (int64,) -> int64
 * (int8,) -> int64
 * (float32,) -> float32
 * (float64,) -> float64
 * (complex64,) -> complex64
 * (complex128,) -> complex128
 * parameterized
[1] During: typing of intrinsic-call at foo.py (4)

File "foo.py", line 4:
def foo(x):
    return [x] + x
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

File: numba/errors.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/errors.py]: Seems like I managed to correct that to another incorrect spelling :) thanks, fixed.

**Comment by stuartarchibald** [comment]
RE https://github.com/numba/numaba/pull/2859#issuecomment-378969770, thanks, I can reproduce, I'll create a fix.

**Comment by stuartarchibald** [comment]
RE: https://github.com/numba/numba/pull/2859#issuecomment-378970218, I can reproduce, this seems to be a Python 2 vs. 3 difference. I'll take a look.

**Comment by stuartarchibald** [comment]
I think I've got to the bottom of the issue, it is pure 2 vs 3, I've made it so it now does this under Python 2.7:
```python
$ ipython
Python 2.7.14 |Anaconda, Inc.| (default, Mar 27 2018, 17:29:31) 
Type "copyright", "credits" or "license" for more information.

IPython 5.6.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]:  from numba import njit

In [2]: @njit
   ...: def foo(x):
   ...:     return [x] + x
   ...: 

In [3]: foo(1)
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-3-7d8d25071659> in <module>()
----> 1 foo(1)

<path>numba/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    338                 raise e
    339             else:
--> 340                 reraise(type(e), e, None)
    341         except errors.UnsupportedError as e:
    342             # Something unsupported is present in the user code, add help info

<path>numba/numba/six.pyc in reraise(tp, value, tb)

TypingError: Failed at nopython (nopython frontend)
Invalid usage of + with parameters (list(int64), int64)
Known signatures:
 * (int64, int64) -> int64
 * (int64, uint64) -> int64
 * (uint64, int64) -> int64
 * (uint64, uint64) -> uint64
 * (float32, float32) -> float32
 * (float64, float64) -> float64
 * (complex64, complex64) -> complex64
 * (complex128, complex128) -> complex128
 * (uint64,) -> uint64
 * (uint16,) -> uint64
 * (uint8,) -> uint64
 * (uint32,) -> uint64
 * (int32,) -> int64
 * (int16,) -> int64
 * (int64,) -> int64
 * (int8,) -> int64
 * (float32,) -> float32
 * (float64,) -> float64
 * (complex64,) -> complex64
 * (complex128,) -> complex128
 * parameterized
[1] During: typing of intrinsic-call at <ipython-input-2-9d8c12fec5f5> (3)

File "<ipython-input-2-9d8c12fec5f5>", line 3:
def foo(x):
    return [x] + x
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```
and if env var `NUMBA_FULL_TRACEBACKS` is set then 
```python
Traceback (most recent call last):
  File "foo23.py", line 10, in <module>
    bar(10)
  File "<path>/numba/numba/dispatcher.py", line 338, in _compile_for_args
    raise e
numba.errors.TypingError: Caused By:
Traceback (most recent call last):
  File "<path>/numba/numba/compiler.py", line 244, in run
    stage()
  File "<path>/numba/numba/compiler.py", line 458, in stage_nopython_frontend
    self.locals)
  File "<path>/numba/numba/compiler.py", line 886, in type_inference_stage
    typemap, restype, calltypes = infer.unify()
  File "<path>/numba/numba/typeinfer.py", line 914, in unify
    retty = self.get_return_type(typdict)
  File "<path>/numba/numba/typeinfer.py", line 994, in get_return_type
    return self._unify_return_types(rettypes)
  File "<path>/numba/numba/typeinfer.py", line 984, in _unify_return_types
    "\n" + "\n".join(problem_str))
TypingError: Can't unify return type from the following types: int64, list(int64)
Return of: IR name '$19.2', type 'int64', location: 
File "foo23.py", line 8:
def bar(x):
    <source elided>
    else:
       return 7
       ^
Return of: IR name '$12.3', type 'list(int64)', location: 
File "foo23.py", line 6:
def bar(x):
    <source elided>
    if x < 3:
       return [x]
       ^

Can't unify return type from the following types: int64, list(int64)
This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

**Comment by sklam** [comment]
A warning about the absence of `colorama` is showing up on import.  I think we shouldn't be verbose about it.

**Comment by stuartarchibald** [comment]
RE: https://github.com/numba/numba/pull/2859#issuecomment-380849247 I can completely remove it? Making it based on `config` will be hard IIRC as there's a circular imports all happening at init time.

**Comment by sklam** [comment]
I think can completely remove the warning.

**Comment by stuartarchibald** [comment]
I manually tested the changes to conda recipes that are now using `run_test.{sh,bat}` in place of `meta.yaml` commands so that env vars can be set on windows (unix is fine, but both are changed for consistency).

**Pull request merged**